src/coprocessor/endpoint.rs

```diff
@@ -293,7 +304,18 @@ impl<E: Engine> Endpoint<E> {
                     .map(|_| (tracker, snapshot))
             })
             .and_then(move |(tracker, snapshot)| {
-                future::result(handler_builder(snapshot, &tracker.req_ctx))
+                let mut builder = handler_builder;
+
+                if let Some(match_version) = tracker.req_ctx.cache_match_version {
```

Comment:
```rust
if tracker.req_ctx.cache_match_version == snapshot.get_data_version() {
    builder = CachedRequestHandler::builder();
}
```

Reply:
we shouldn't return cache if they are two `None`s .